src/models/state/wallet/wallet_state.rs

```diff
@@ -225,13 +228,38 @@ impl WalletState {
         let rusty_wallet_database = RustyWalletDatabase::connect(wallet_db).await;
         let sync_label = rusty_wallet_database.get_sync_label().await;
 
+        // generate and cache all used generation keys
+        let mut known_generation_keys =
+            (0..rusty_wallet_database.get_generation_key_counter().await)
+                .map(|idx| wallet_secret.nth_generation_spending_key(idx).into())
+                .collect_vec();
+
+        // this is a hack to appease existing tests that bypass
+        // WalletState::next_unused_generation_spending_key() by
+        // calling WalletSecret::nth_generation_spending_key_for_tests().
```

Comment:
Yeah, this is a bit intricate.  I will unpack... 

So the initial issue I noticed is that many tests call WalletSecret::nth_generation_spending_key_for_tests(0) to directly obtain the 0-th key.  This bypasses the logic in next_unused_generation_spending_key() that increments the wallet's derivation counter and adds the key to known_keys list.

I started a sub-branch to update the tests, but it is not always feasible.    Some of the tests do not involve GlobalState or even WalletState.   Such tests are not really a problem though as they are not testing things that can be affected by not using the counter.

Then I realized that the larger issue is that the application logic uses the 0th generation key for coinbase and pre-mine purposes.   It is assumed to exist (be known).  But if we initially add it to the known-keys list (this hack) and subsequently call next_unused_spending_key() then it is added again, resulting in a dup.  I worked around dups by using a HashSet for known-keys, but that adds some hashing overhead and feels like a bandaid.

I determined that a better/correct fix is to call next_unused_spending_key() inside WalletState::new_from_wallet_secret() immediately after WalletState is created.   

That works correctly with no dups.  The downside is that it results in some tests generating different proofs than we have on the proof server because when the test calls next_unused_spending_key() it obtains the key at index 1 instead of index 0 as it used to.

But it seems objectively correct. 

*Given that coinbase rewards are always paid to key 0 then index 0 should be reserved for coinbase and user-level tx should start at index 1.*  

That's really the important thing to realize here.   and pre-mine rewards are a special case of coinbase rewards.

Then we must ask: why are coinbase rewards always paid to index 0?  Why not derive a addr for each coinbase?    The code presently does it this way because we can avoid acquiring a global-state write-lock (to mutate wallet state) when constructing block template(s) by always using the first key.  There doesn't seem to be a privacy concern because the coinbase ExpectedUtxo is kept offchain by the miner.  There is a comment in mine_loop.rs to this effect.

I should note that any wallet that is not mining and not a pre-mine recipient also has their 0th address skipped for receiving purposes.   This seems fine/correct because any such wallet could begin mining at some point in the future and receive a coinbase reward at key 0.  hence it is *reserved*.

Assuming we go forward with this fix/approach, I may need some assistance getting new proofs onto the proof server(s) since I don't have access.  It's kind of a small detail, but a headache practically for me when testing, pushing to PR, etc.

btw, concretely the discussed approach is to replace the above-referenced "hack" with the following after WalletState is instantiated:

```
        // key 0 is reserved for coinbase and pre-mine.
        // this ensures derivation-index=0 key is known to wallet, so coinbase and premine claiming works.
        // We call it here to avoid acquiring write-lock in mining task when generating tx template.
        // wallets start at index 1 for all other tx.
        if wallet_state.known_generation_keys.is_empty() {
            let _ = wallet_state.next_unused_spending_key(KeyType::Generation).await;
        }
```

Reply:
> `   // note: it is Ok to always use the same key here because:`
> `    //  1. if we find a block, the utxo will go to our wallet`
> `    //     and notification occurs offchain, so there is no privacy issue.`
> `    //  2. if we were to derive a new addr for each block then we would`
> `    //     have large gaps since an address only receives funds when`
> `    //     we actually win the mining lottery.`
> `    //  3. also this way we do not have to modify global/wallet state.`
> 
> However @aszepieniec I would like to make sure you agree with the privacy assessment and overall rationale, that it is ok to always mine to the 0 key.

Short answer: yes I agree. :heavy_check_mark: 

Long answer: but it's not obvious that there *isn't* a privacy hazard. In particular, I was afraid that the coinbase UTXO's addition record was going to be identical if they are from the same miner in the same halving epoch. Indeed, the UTXO is identical and the receiver digest is too. But as it turns out, the sender randomness takes the block height into account so we are good.

There is an edge case though, which is when a fork happens to a shorter chain (fewer blocks but more proof-of-work) and the same miner mines block *k*. Well then it will be clear to the outside world that it's the same miner who composed block *k*. I think that is an acceptable compromise.

Reply:
thanks for adding these considerations.    this thread seems to be resolved.